packages/dds/sequence/src/defaultMap.ts

```diff
@@ -0,0 +1,451 @@
+ * Licensed under the MIT License.
+ */
+
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { IFluidSerializer, makeHandlesSerializable, parseHandles, ValueType } from "@fluidframework/shared-object-base";
+import { assert, TypedEventEmitter } from "@fluidframework/common-utils";
+import {
+    ILocalValue,
+    LocalValueMaker,
+    makeSerializable,
+    ValueTypeLocalValue,
+} from "./localValues";
+import {
+    ISerializableValue,
+    ISerializedValue,
+    IValueChanged,
+    IValueOpEmitter,
+    IValueType,
+    IValueTypeOperationValue,
+    ISharedDefaultMapEvents,
+} from "./defaultMapInterfaces";
+
+/**
+ * Defines the means to process and submit a given op on a map.
+ */
+interface IMapMessageHandler {
+    /**
+     * Apply the given operation.
+     * @param op - The map operation to apply
+     * @param local - Whether the message originated from the local client
+     * @param message - The full message. Not provided for stashed ops.
+     * @param localOpMetadata - For local client messages, this is the metadata that was submitted with the message.
+     * For messages from a remote client, this will be undefined.
+     */
+    process(
+        op: IMapOperation,
+        local: boolean,
+        message: ISequencedDocumentMessage | undefined,
+        localOpMetadata: IMapMessageLocalMetadata,
+    ): void;
+
+    /**
+     * Communicate the operation to remote clients.
+     * @param op - The map operation to submit
+     */
+    submit(op: IMapOperation): void;
+
+    getStashedOpLocalMetadata(op: IMapOperation): unknown;
+}
+
+export interface IMapMessageLocalMetadata {
+    lastProcessedSeq: number;
+}
+
+/**
+ * Describes an operation specific to a value type.
+ */
+export interface IMapValueTypeOperation {
+    /**
+     * String identifier of the operation type.
+     */
+    type: "act";
+
+    /**
+     * Map key being modified.
+     */
+    key: string;
+
+    /**
+     * Value of the operation, specific to the value type.
+     * @alpha
+     */
+    value: IValueTypeOperationValue;
+}
+
+/**
+ * Description of a map delta operation
+ */
+export type IMapOperation = IMapValueTypeOperation;
+
+/**
+ * Defines the in-memory object structure to be used for the conversion to/from serialized.
+ * Directly used in JSON.stringify, direct result from JSON.parse
+ */
+export interface IMapDataObjectSerializable {
+    [key: string]: ISerializableValue;
+}
+
+export interface IMapDataObjectSerialized {
+    [key: string]: ISerializedValue;
+}
+
+/**
+ * A DefaultMap is a map-like distributed data structure, supporting operations on values stored by
+ * string key locations.
+ *
+ * Creation of values is implicit on access (either via `get` or a remote op application referring to
+ * a collection that wasn't previously known)
+ */
+export class DefaultMap<T> {
```

Comment:
is there a need for this to be generic? i was thinking we would just build something specialized to intervals

Reply:
I'm cool with collapsing the abstractions entirely. Initially only took the less drastic step of genericizing because things were really excessively typed as `any`.

Reply:
yeah. i don't think we need any more it, pun intended :) i really just want this simplified, so it's easier to maintain

Reply:
Right now we use this with `T` parameterized as `IntervalCollection<Interval>` and `IntervalCollection<SequenceInterval>` (in `SharedIntervalCollection` and `Sequence` respectively). IIRC the former is deprecated, but I think as long as the code is around it makes sense to keep this generic, i'd prefer it to having to cast the result of `.get()` in those two DDSes.